display/renderbasic/src/main/java/org/openscience/cdk/renderer/elements/Bounds.java

```diff
@@ -59,6 +73,114 @@ public Bounds(double x1, double y1, double x2, double y2) {
+    public Bounds() {
+        this(+Double.MAX_VALUE, +Double.MAX_VALUE,
+             -Double.MAX_VALUE, -Double.MAX_VALUE);
+    }
+
+    /**
+     * An bounding box around the specified element.
+     */
+    public Bounds(IRenderingElement element) {
+        this();
+        add(element);
+    }
+
+    /**
+     * Add the specified element bounds.
+     */
+    public void add(IRenderingElement element) {
+        elements.add(element);
+        traverse(element);
+    }
+
+    /**
+     * Ensure the point x,y is included in the bounding box.
+     *
+     * @param x x-coordinate
+     * @param y y-coordinate
+     */
+    public void add(double x, double y) {
+        if (x < minX) minX = x;
+        if (y < minY) minY = y;
+        if (x > maxX) maxX = x;
+        if (y > maxY) maxY = y;
+    }
+
+    /**
+     * Add one bounds to another.
+     *
+     * @param bounds other bounds
+     */
+    public void add(Bounds bounds) {
+        if (bounds.minX < minX) minX = bounds.minX;
+        if (bounds.minY < minY) minY = bounds.minY;
+        if (bounds.maxX > maxX) maxX = bounds.maxX;
+        if (bounds.maxY > maxY) maxY = bounds.maxY;
+    }
+
+    /**
+     * Add the provided general path to the bounding box.
+     *
+     * @param path general path
+     */
+    public void add(GeneralPath path) {
+        double[] points = new double[6];
+        for (org.openscience.cdk.renderer.elements.path.PathElement element : path.elements) {
+            element.points(points);
+            switch (element.type()) {
+                case MoveTo:
+                case LineTo:
+                    add(points[0], points[1]);
+                    break;
+                case QuadTo:
+                    add(points[2], points[3]);
+                    break;
+                case CubicTo:
+                    add(points[4], points[5]);
+                    break;
+            }
+        }
+    }
+
+    private void traverse(IRenderingElement newElement) {
+        Deque<IRenderingElement> stack = new ArrayDeque<>();
+        stack.push(newElement);
+        while (!stack.isEmpty()) {
+            final IRenderingElement element = stack.poll();
+            if (element instanceof Bounds) {
+                add((Bounds) element);
+            } else if (element instanceof GeneralPath) {
+                add((GeneralPath) element);
+            } else if (element instanceof LineElement) {
+                LineElement lineElem = (LineElement) element;
+                add(lineElem.firstPointX, lineElem.firstPointY);
+                add(lineElem.secondPointX, lineElem.secondPointY);
+            } else if (element instanceof ElementGroup) {
+                for (IRenderingElement child : (ElementGroup) element)
+                    stack.add(child);
+            } else {
+                // ignored from bounds calculation, we don't really
+                // care but log we skipped it
+                LoggingToolFactory.createLoggingTool(Bounds.class)
```

Comment:
Is this going to produce a warning for every leaf that is not a line or a general path? That seems quite noisy...